risc0/zkvm/src/prove/mod.rs

```diff
@@ -229,16 +281,87 @@ impl Segment {
         log::debug!("Globals: {:?}", OutBuffer(out_slice).tree(&LAYOUT));
         let out = hal.copy_from_elem("out", &adapter.get_io().as_slice());
 
-        let seal = prover.finalize(&[&mix, &out], eval);
+        let seal = prover.finalize(&[&mix, &out], eval.as_ref());
 
         let receipt = SegmentReceipt { seal };
         let hal = CpuVerifyHal::<_, H::HashSuite, _>::new(&crate::CIRCUIT);
-        receipt
-            .verify_with_hal(&hal)
-            .map_err(|err| anyhow!("Verification error: {err}"))?;
+        receipt.verify_with_hal(&hal)?;
 
         Ok(receipt)
     }
+}
+
+fn provers() -> HashMap<String, Arc<dyn Prover>> {
+    let mut table: HashMap<String, Arc<dyn Prover>> = HashMap::new();
+    {
+        let prover = Arc::new(LocalProver::new("cpu", cpu::sha256_hal_eval()));
+        table.insert("cpu".to_string(), prover.clone());
+        table.insert("$default".to_string(), prover);
+
+        let prover = Arc::new(LocalProver::new("cpu:poseidon", cpu::poseidon_hal_eval()));
+        table.insert("cpu:poseidon".to_string(), prover.clone());
+        table.insert("$poseidon".to_string(), prover);
+    }
+    #[cfg(feature = "cuda")]
+    {
+        let prover = Arc::new(LocalProver::new("cuda", cuda::sha256_hal_eval()));
+        table.insert("cuda".to_string(), prover.clone());
+        table.insert("$gpu".to_string(), prover.clone());
+        table.insert("$default".to_string(), prover);
+
+        let prover = Arc::new(LocalProver::new("cuda:poseidon", cuda::poseidon_hal_eval()));
+        table.insert("cuda:poseidon".to_string(), prover.clone());
+        table.insert("$poseidon".to_string(), prover);
+    }
+    #[cfg(feature = "metal")]
+    {
+        let prover = Arc::new(LocalProver::new("metal", metal::sha256_hal_eval()));
+        table.insert("metal".to_string(), prover.clone());
+        table.insert("$gpu".to_string(), prover.clone());
+        table.insert("$default".to_string(), prover);
+
+        let prover = Arc::new(LocalProver::new(
+            "metal:poseidon",
+            metal::poseidon_hal_eval(),
+        ));
+        table.insert("metal:poseidon".to_string(), prover.clone());
+        table.insert("$poseidon".to_string(), prover);
+    }
+    table
+}
+
+/// Return a default [Prover] based on environment variables, falling back to a
+/// default CPU-based prover.
+pub fn default_prover() -> Arc<dyn Prover> {
+    let provers = provers();
+    if let Ok(requested) = std::env::var("RISC0_PROVER") {
```

Comment:
We should probably document this in a more expanded rustdoc str and potentially the README. I think we are now getting to the point where we should consider adding all the env vars to list with explanations of each.

Reply:
This is a great idea -- creating an issue for it.